File: android/src/main/java/com/reactnativeandroidauto/AndroidAutoModule.java
package com.reactnativeandroidauto;

import android.content.Intent;
import android.net.Uri;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import androidx.activity.OnBackPressedCallback;
import androidx.car.app.CarContext;
import androidx.car.app.CarToast;
import androidx.car.app.ScreenManager;
import androidx.car.app.model.Template;

import com.facebook.react.bridge.Callback;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableNativeMap;
import com.facebook.react.module.annotations.ReactModule;
import com.facebook.react.modules.core.DeviceEventManagerModule;
import com.facebook.react.modules.debug.DevSettingsModule;

import java.util.WeakHashMap;

@ReactModule(name = AndroidAutoModule.MODULE_NAME)
public class AndroidAutoModule extends ReactContextBaseJavaModule {
    static final String MODULE_NAME = "CarModule";

    private static ReactApplicationContext mReactContext;
    private CarContext mCarContext;
    private CarScreen mCurrentCarScreen;
    private ScreenManager mScreenManager;

    private WeakHashMap<String, CarScreen> carScreens;
    private WeakHashMap<CarScreen, ReactCarRenderContext> reactCarRenderContextMap;

    public String getName() {
        return MODULE_NAME;
    }

    private Handler handler;

    AndroidAutoModule(ReactApplicationContext context) {
        super(context);
        carScreens = new WeakHashMap();
        reactCarRenderContextMap = new WeakHashMap();
        handler = new Handler(Looper.getMainLooper());

        handler.postDelayed(() -> {
            Log.d("TODO", "sending event to the react side");
            sendEvent("autoTesting", new WritableNativeMap());
        }, 5000);
        mReactContext = context;
    }

    @ReactMethod
    public void invalidate(String name) {
        handler.post(() -> {
            CarScreen screen = getScreen(name);
            if (screen == mScreenManager.getTop()) {
                screen.invalidate();
            }
        });
    }

    @ReactMethod
    public void setTemplate(String name, ReadableMap renderMap, Callback callback) {
        ReactCarRenderContext reactCarRenderContext = new ReactCarRenderContext();
        CarScreen screen = getScreen(name);
        if (screen == null) {
            screen = mCurrentCarScreen;
            Log.d("AUTO", "Screen " + name + " not found!");
        }

        reactCarRenderContext.setEventCallback(callback).setScreenMarker(screen.getMarker());

        Template template = parseTemplate(renderMap, reactCarRenderContext);
        reactCarRenderContextMap.remove(screen);
        reactCarRenderContextMap.put(screen, reactCarRenderContext);

        screen.setTemplate(template);
    }

    @ReactMethod
    public void pushScreen(String name, ReadableMap renderMap, Callback callback) {
        ReactCarRenderContext reactCarRenderContext = new ReactCarRenderContext();
        reactCarRenderContext.setEventCallback(callback).setScreenMarker(name);

        Template template = parseTemplate(renderMap, reactCarRenderContext);

        CarScreen screen = new CarScreen(mCarContext);
        reactCarRenderContextMap.remove(screen);
        reactCarRenderContextMap.put(screen, reactCarRenderContext);

        screen.setMarker(name);
        screen.setTemplate(template);
        carScreens.put(name, screen);
        mCurrentCarScreen = screen;
        mScreenManager.push(screen);
    }

    @ReactMethod
    public void popScreen() {
        mScreenManager.pop();
        removeScreen(mCurrentCarScreen);
        mCurrentCarScreen = (CarScreen) mScreenManager.getTop();
    }

    @ReactMethod
    public void mapNavigate(String address) {
        mCarContext.startCarApp(new Intent(CarContext.ACTION_NAVIGATE, Uri.parse("geo:0,0?q=" + address)));
    }

    @ReactMethod
    public void toast(String text, int duration) {
        CarToast.makeText(mCarContext, text, duration).show();
    }

    @ReactMethod
    public void reload() {
        DevSettingsModule devSettingsModule = mReactContext.getNativeModule(DevSettingsModule.class);
        if (devSettingsModule != null) {
            devSettingsModule.reload();
        }
    }

    @ReactMethod
    public void finishCarApp() {
        mCarContext.finishCarApp();
    }

    @ReactMethod
    public void setEventCallback(String name, Callback callback) {
        CarScreen screen = (CarScreen) getScreen(name);

        Log.d("AUTO", "Set callback 1 for " + name);
        if (screen == null) {
            return;
        }

        Log.d("AUTO", "Set callback 2 for " + name);
        ReactCarRenderContext reactCarRenderContext = reactCarRenderContextMap.get(screen);

        if (reactCarRenderContext == null) {
            return;
        }

        Log.d("AUTO", "Set callback 3 for " + name);
        reactCarRenderContext.setEventCallback(callback);
    }

    public void setCarContext(CarContext carContext, CarScreen currentCarScreen) {
        mCarContext = carContext;
        mCurrentCarScreen = currentCarScreen;
        mScreenManager = currentCarScreen.getScreenManager();
        carScreens.put("root", mCurrentCarScreen);

        OnBackPressedCallback callback = new OnBackPressedCallback(true) {
            @Override
            public void handleOnBackPressed() {
                Log.d("AUTO", "Back button pressed");
                sendEvent("android_auto:back_button", new WritableNativeMap());
            }
        };

        carContext.getOnBackPressedDispatcher().addCallback(callback);

        handler.postDelayed(() -> {
            Log.d("TODO", "sending event to the react side");
            sendEvent("android_auto:ready", new WritableNativeMap());
            sendEvent("autoTesting", new WritableNativeMap());
        }, 1000);
    }

    private Template parseTemplate(ReadableMap renderMap, ReactCarRenderContext reactCarRenderContext) {
        TemplateParser templateParser = new TemplateParser(reactCarRenderContext);
        return templateParser.parseTemplate(renderMap);
    }

    private CarScreen getScreen(String name) {
        return carScreens.get(name);
    }

    private void removeScreen(CarScreen screen) {
        WritableNativeMap params = new WritableNativeMap();
        params.putString("screen", screen.getMarker());

        sendEvent("android_auto:remove_screen", params);

        carScreens.values().remove(screen);
    }

    private void sendEvent(String eventName, Object params) {
        mReactContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit(eventName, params);
    }
}

File: android/src/main/java/com/reactnativeandroidauto/AndroidAutoModule.kt
package com.reactnativeandroidauto

import android.content.Intent
import android.net.Uri
import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.activity.OnBackPressedCallback
import androidx.car.app.CarContext
import androidx.car.app.CarToast
import androidx.car.app.ScreenManager
import androidx.car.app.model.Template
import com.facebook.react.bridge.*
import com.facebook.react.module.annotations.ReactModule
import com.facebook.react.modules.core.DeviceEventManagerModule.RCTDeviceEventEmitter
import com.facebook.react.modules.debug.DevSettingsModule
import com.reactnativeandroidauto.template.TemplateFactory
import java.util.*

@ReactModule(name = AndroidAutoModule.MODULE_NAME)
class AndroidAutoModule internal constructor(private val reactContext: ReactApplicationContext) :
  ReactContextBaseJavaModule(reactContext) {
  private lateinit var carContext: CarContext
  private var currentCarScreen: CarScreen? = null
  private var screenManager: ScreenManager? = null
  private val carScreens: WeakHashMap<String, CarScreen> = WeakHashMap()
  private val reactCarRenderContextMap: WeakHashMap<CarScreen, ReactCarRenderContext> =
    WeakHashMap()
  private val handler: Handler = Handler(Looper.getMainLooper())

  override fun getName(): String {
    return MODULE_NAME
  }

  @ReactMethod
  fun invalidate(name: String) {
    handler.post {
      val screen = getScreen(name)
      if (screen === screenManager!!.top) {
        screen.invalidate()
      }
    }
  }

  @ReactMethod
  fun setTemplate(name: String, renderMap: ReadableMap, callback: Callback?) {
    val screen = getScreen(name)
    if (screen == null) {
      Log.d(TAG, "Screen $name not found!")
      return
    }
    val reactCarRenderContext = ReactCarRenderContext(screen.marker!!, callback)
    val template = parseTemplate(renderMap, reactCarRenderContext)
    reactCarRenderContextMap.remove(screen)
    reactCarRenderContextMap[screen] = reactCarRenderContext
    screen.setTemplate(template, renderMap)
  }

  @ReactMethod
  fun pushScreen(name: String?, renderMap: ReadableMap, callback: Callback?) {
    handler.post {
      val reactCarRenderContext = ReactCarRenderContext(name!!, callback)
      val screen = CarScreen(carContext)
      screenManager = screen.screenManager
      reactCarRenderContextMap.remove(screen)
      reactCarRenderContextMap[screen] = reactCarRenderContext
      screen.marker = name
      val template = parseTemplate(renderMap, reactCarRenderContext)
      screen.setTemplate(template, renderMap)
      carScreens[name] = screen
      currentCarScreen = screen
      screenManager!!.push(screen)
    }
  }

  @ReactMethod
  fun popScreen() {
    handler.post {
      screenManager!!.pop()
      removeScreen(currentCarScreen)
      currentCarScreen = screenManager!!.top as CarScreen
    }
  }

  @ReactMethod
  fun mapNavigate(address: String) {
    carContext.startCarApp(
      Intent(
        CarContext.ACTION_NAVIGATE,
        Uri.parse("geo:0,0?q=$address")
      )
    )
  }

  @ReactMethod
  fun toast(text: String, duration: Int) {
    CarToast.makeText(carContext, text, duration).show()
  }

  @ReactMethod
  fun reload() {
    val devSettingsModule = reactContext.getNativeModule(
      DevSettingsModule::class.java
    )
    devSettingsModule?.reload()
  }

  @ReactMethod
  fun finishCarApp() {
    carContext.finishCarApp()
  }

  @ReactMethod
  fun setEventCallback(name: String, callback: Callback?) {
    val screen = getScreen(name) ?: return
    val reactCarRenderContext = reactCarRenderContextMap[screen] ?: return
    reactCarRenderContext.eventCallback = callback
  }

  fun setCarContext(carContext: CarContext, currentCarScreen: CarScreen) {
    this.carContext = carContext
    this.currentCarScreen = currentCarScreen
    screenManager = currentCarScreen.screenManager
    carScreens["root"] = this.currentCarScreen
    val callback: OnBackPressedCallback = object : OnBackPressedCallback(true) {
      override fun handleOnBackPressed() {
        Log.d(TAG, "Back button pressed")
        sendEvent("android_auto:back_button", WritableNativeMap())
      }
    }
    carContext.onBackPressedDispatcher.addCallback(callback)
    sendEvent("android_auto:ready", WritableNativeMap())
  }

  private fun parseTemplate(
    renderMap: ReadableMap,
    reactCarRenderContext: ReactCarRenderContext
  ): Template {
    val factory = TemplateFactory(carContext, reactCarRenderContext)
    return factory.getTemplate(renderMap)
  }

  private fun getScreen(name: String): CarScreen? {
    return carScreens[name]
  }

  private fun removeScreen(screen: CarScreen?) {
    val params = WritableNativeMap()
    params.putString("screen", screen!!.marker)
    sendEvent("android_auto:remove_screen", params)
    carScreens.values.remove(screen)
  }

  private fun sendEvent(eventName: String, params: Any) {
    reactContext
      .getJSModule(RCTDeviceEventEmitter::class.java)
      .emit(eventName, params)
  }

  companion object {
    const val MODULE_NAME = "CarModule"
    const val TAG = "AndroidAutoModule"
  }

}
